imports/wasi_snapshot_preview1/poll.go

```diff
@@ -233,10 +236,60 @@ func processClockEvent(inBuf []byte) (time.Duration, sys.Errno) {
 
 // writeEvent writes the event corresponding to the processed subscription.
 // https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-event-struct
-func writeEvent(outBuf []byte, evt *event) {
-	copy(outBuf[evt.outOffset:], evt.userData) // userdata
-	outBuf[evt.outOffset+8] = byte(evt.errno)  // uint16, but safe as < 255
-	outBuf[evt.outOffset+9] = 0
-	le.PutUint32(outBuf[evt.outOffset+10:], uint32(evt.eventType))
+func writeEvent(outBuf []byte, evt *pollEvent) {
+	copy(outBuf, evt.userData)  // userdata
+	outBuf[8] = byte(evt.errno) // uint16, but safe as < 255
+	outBuf[9] = 0
+	le.PutUint32(outBuf[10:], uint32(evt.eventType))
 	// TODO: When FD events are supported, write outOffset+16
 }
+
+// closeChAfter closes a channel after the given timeout.
+// It is similar to time.After but it uses sysCtx.Nanosleep.
+func closeChAfter(sysCtx *internalsys.Context, timeout time.Duration, timeoutCh chan struct{}) {
+	sysCtx.Nanosleep(int64(timeout))
+	close(timeoutCh)
+}
+
+// pollFileEventsOnce invokes Poll on each sys.FileEntry in the given slice
+// and writes back the result to outBuf for each file reported "ready";
+// i.e., when Poll() returns true, and no error.
+func pollFileEventsOnce(evts []*filePollEvent, outBuf []byte) (n uint32, errno sys.Errno) {
+	// For simplicity, we assume that there are no multiple subscriptions for the same file.
+	for _, e := range evts {
+		isReady, errno := e.f.File.Poll(sys.POLLIN, 0)
+		if errno != 0 {
+			return 0, errno
+		}
+		if isReady {
+			e.e.errno = 0
+			writeEvent(outBuf[n*32:], e.e)
+			n++
+		}
+	}
+	return
+}
+
+// pollFileEventsUntil repeatedly invokes pollFileEventsOnce until the given timeout is reached.
+// The poll interval is currently fixed at 100 millis.
+func pollFileEventsUntil(sysCtx *internalsys.Context, timeout time.Duration, blockingSubs []*filePollEvent, outBuf []byte) (n uint32, errno sys.Errno) {
+	timeoutCh := make(chan struct{}, 1)
+	go closeChAfter(sysCtx, timeout, timeoutCh)
+
+	pollInterval := 100 * time.Millisecond
+	ticker := time.NewTicker(pollInterval)
```

Comment:
fyi closeChAfter we are intentionally using the context clock, but this will use a real one..

Reply:
whoops!